atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1862,12 +1871,51 @@ public void commit() {
     @Override
     public void commit(TransactionService transactionService) {
         try {
+            runPreCommitCallbacks();
             commitWithoutCallbacks(transactionService);
         } finally {
             close();
         }
     }
 
+    @Override
+    public void preCommit(
+            TimestampLeaseName leaseName, int numLeasedTimestamps, Consumer<LongSupplier> preCommitAction) {
+        Preconditions.checkArgument(numLeasedTimestamps > 0, "Need to request a non-empty number of timestamp leases");
+        preCommitActions.addPreCommitAction(leaseName, numLeasedTimestamps, preCommitAction);
+    }
+
+    @ReviewedRestrictedApiUsage
+    private void runPreCommitCallbacks() {
+        Map<TimestampLeaseName, PerLeaseActions> actions = preCommitActions.getActions();
+        if (actions.isEmpty()) {
+            return;
+        }
+
+        TimestampLeaseResults timestampLeaseResults =
+                timelockService.acquireTimestampLeases(Maps.transformValues(actions, action -> action.timestampCount));
+        transactionLocksManager.registerLock(timestampLeaseResults.lock());
+
+        actions.forEach((timestampLeaseName, perLeaseActions) -> {
+            perLeaseActions.preCommitActions.forEach(preCommitAction -> {
+                LongSupplier leasedTimestamps;
+                if (preCommitAction.timestampCount == 0) {
+                    leasedTimestamps = () -> {
+                        throw new SafeRuntimeException(
+                                "Cannot fetch leased timestamps since pre-commit action requested 0 leased timestamps",
+                                SafeArg.of("leaseName", timestampLeaseName));
+                    };
```

Comment:
question: we have the 0 case for some internal usage? It seems that to add an action via `preCommit`, you're not allowed for this to be zero

Reply:
We have theoretical ones, but I'll expose it as a new public method in `Transaction` with simpler signature - only `preCommit(Runnable)`.